File: sdks/typescript/src/hyperloglog.ts
import { SynapClient } from './client';
import type { CommandOptions, HyperLogLogStats } from './types';

function encodeElement(element: string | Uint8Array | number[]): number[] {
  if (typeof element === 'string') {
    return Array.from(new TextEncoder().encode(element));
  }

  if (element instanceof Uint8Array) {
    return Array.from(element);
  }

  if (Array.isArray(element)) {
    return element;
  }

  throw new TypeError('Unsupported element type for HyperLogLog. Use string, Uint8Array, or number[].');
}

interface PfAddResponse {
  added?: number;
}

interface PfCountResponse {
  count?: number;
}

interface PfMergeResponse {
  count?: number;
}

/**
 * HyperLogLog manager for probabilistic cardinality estimation.
 */
export class HyperLogLogManager {
  constructor(private readonly client: SynapClient) {}

  private buildPayload(
    options: CommandOptions | undefined,
    extra: Record<string, unknown>
  ): Record<string, unknown> {
    const payload: Record<string, unknown> = { ...extra };

    if (options?.clientId) {
      payload.client_id = options.clientId;
    }

    return payload;
  }

  /**
   * Add elements to a HyperLogLog structure (PFADD)
   */
  async pfadd(
    key: string,
    elements: Array<string | Uint8Array | number[]>,
    options?: CommandOptions
  ): Promise<number> {
    if (!elements.length) {
      return 0;
    }

    const encoded = elements.map(encodeElement);
    const response = await this.client.sendCommand<PfAddResponse>(
      'hyperloglog.pfadd',
      this.buildPayload(options, {
        key,
        elements: encoded,
      })
    );

    return response.added ?? 0;
  }

  /**
   * Estimate cardinality of a HyperLogLog structure (PFCOUNT)
   */
  async pfcount(key: string, options?: CommandOptions): Promise<number> {
    const response = await this.client.sendCommand<PfCountResponse>(
      'hyperloglog.pfcount',
      this.buildPayload(options, { key })
    );

    return response.count ?? 0;
  }

  /**
   * Merge multiple HyperLogLog structures (PFMERGE)
   */
  async pfmerge(
    destination: string,
    sources: string[],
    options?: CommandOptions
  ): Promise<number> {
    const response = await this.client.sendCommand<PfMergeResponse>(
      'hyperloglog.pfmerge',
      this.buildPayload(options, {
        destination,
        sources,
      })
    );

    return response.count ?? 0;
  }

  /**
   * Retrieve HyperLogLog statistics
   */
  async stats(options?: CommandOptions): Promise<HyperLogLogStats> {
    const response = await this.client.sendCommand<HyperLogLogStats>(
      'hyperloglog.stats',
      this.buildPayload(options, {})
    );

    return response;
  }
}

File: sdks/rust/src/hyperloglog.rs
//! HyperLogLog operations (PFADD/PFCOUNT/PFMERGE)

use crate::client::SynapClient;
use crate::error::Result;
use crate::types::HyperLogLogStats;
use serde_json::json;

#[derive(Clone)]
pub struct HyperLogLogManager {
    client: SynapClient,
}

impl HyperLogLogManager {
    pub(crate) fn new(client: SynapClient) -> Self {
        Self { client }
    }

    /// Add elements to a HyperLogLog structure (PFADD)
    pub async fn pfadd<I, T>(&self, key: &str, elements: I) -> Result<usize>
    where
        I: IntoIterator<Item = T>,
        T: AsRef<[u8]>,
    {
        let encoded: Vec<Vec<u8>> = elements
            .into_iter()
            .map(|el| el.as_ref().to_vec())
            .collect();
        if encoded.is_empty() {
            return Ok(0);
        }

        let payload = json!({
            "key": key,
            "elements": encoded,
        });

        let response = self
            .client
            .send_command("hyperloglog.pfadd", payload)
            .await?;
        Ok(response["added"].as_u64().unwrap_or(0) as usize)
    }

    /// Estimate cardinality of a HyperLogLog structure (PFCOUNT)
    pub async fn pfcount(&self, key: &str) -> Result<u64> {
        let payload = json!({"key": key});
        let response = self
            .client
            .send_command("hyperloglog.pfcount", payload)
            .await?;
        Ok(response["count"].as_u64().unwrap_or(0))
    }

    /// Merge multiple HyperLogLog structures into destination (PFMERGE)
    pub async fn pfmerge<S>(&self, destination: &str, sources: &[S]) -> Result<u64>
    where
        S: AsRef<str>,
    {
        let payload = json!({
            "destination": destination,
            "sources": sources.iter().map(|s| s.as_ref()).collect::<Vec<_>>(),
        });
        let response = self
            .client
            .send_command("hyperloglog.pfmerge", payload)
            .await?;
        Ok(response["count"].as_u64().unwrap_or(0))
    }

    /// Retrieve HyperLogLog statistics
    pub async fn stats(&self) -> Result<HyperLogLogStats> {
        let response = self
            .client
            .send_command("hyperloglog.stats", json!({}))
            .await?;
        let stats: HyperLogLogStats = serde_json::from_value(response)?;
        Ok(stats)
    }
}
